pioneer/packages/joy-forum/src/ViewThread.tsx

```diff
@@ -124,6 +125,78 @@ type ViewThreadProps = ApiProps & InnerViewThreadProps & {
   nextPostId?: ThreadId;
 };
 
+const POSTS_THREAD_MAP_CACHE_KEY = 'postsThreadMap';
+
+async function refreshPostsInThreadCache (nextPostId: PostId, api: ApiPromise) {
+  const newId = (id: number | BN) => api.createType('PostId', id);
+  const apiCalls: Promise<Post>[] = [];
+  let id = newId(1);
+
+  let mapPostToThread = getPostsIdsInThreadCache();
+  const nextTreadId = await api.query.forum.nextThreadId() as ThreadId;
+
+  if (mapPostToThread.size >= Number(nextTreadId.toString())) { // invalid cache
+    mapPostToThread = new Map<number, number[]>();
+  }
+
+  if (mapPostToThread.size > 0) {
+    let lastPostIdInCache = 0;
+
+    for (const postIds of mapPostToThread.values()) {
+      const maxPostIdInThread = postIds.reduce((a, b) => a > b ? a : b);
+
+      if (maxPostIdInThread > lastPostIdInCache) {
+        lastPostIdInCache = maxPostIdInThread;
+      }
+    }
+
+    id = newId(lastPostIdInCache);
+    const lastPost = await api.query.forum.postById(lastPostIdInCache) as Post;
+
+    if (lastPost) {
+      const postsInThread = mapPostToThread.get(Number(lastPost.thread_id.toString()));
+
+      if (!postsInThread || !postsInThread.includes(lastPostIdInCache)) { // cache doesn't match the data in chain
+        mapPostToThread = new Map<number, number[]>();
+      }
+    } else {
+      mapPostToThread = new Map<number, number[]>();
+    }
+  }
+
+  while (nextPostId.gt(id)) {
+    apiCalls.push(api.query.forum.postById(id) as Promise<Post>);
+    id = newId(id.add(newId(1)));
+  }
+
+  const newPosts = await Promise.all<Post>(apiCalls);
+
+  for (const post of newPosts) {
```

Comment:
I think this part can be optimized a little bit. Currently, if I'm not mistaken the time complexity would be something along the lines of `O(number_of_threads * number_of_posts_in_that_thread^2)`.

Now I'm not sure if it would really make that much of a difference since there aren't going to be too many new posts usually after caching, so I'll leave it up to you whether or not you want to change this.

Reply:
Not sure how this could be improved :thinking: 
Maybe grouping by posts per thread first and then updating the map? :thinking: